### PR TITLE
[IOTDB-1626]TsFile API supports write and register on aligned timeseries

### PR DESCRIPTION
**Pull request by choubenson**
Supporting write/register vector in TsFile API. 
Give four user examples of using TsFile write API.
Upgrade the sequenceRead Tool which can read aligned or nonAligned timeseries in TsFile sequentially.

### CODE REVIEW

File: example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteAlignedWithTSRecord.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  public static void main(String[] args) throws IOException {
+    File f = FSFactoryProducer.getFSFactory().getFile("alignedRecord.tsfile");
+    if (f.exists() && !f.delete()) {
+      throw new RuntimeException("can not delete " + f.getAbsolutePath());
+    }
+    try (TsFileWriter tsFileWriter = new TsFileWriter(f)) {
+      List<UnaryMeasurementSchema> measurementSchemas = new ArrayList<>();
+      measurementSchemas.add(new UnaryMeasurementSchema("s1", TSDataType.INT64, TSEncoding.RLE));
+      measurementSchemas.add(new UnaryMeasurementSchema("s2", TSDataType.INT64, TSEncoding.RLE));
+      measurementSchemas.add(new UnaryMeasurementSchema("s3", TSDataType.INT64, TSEncoding.RLE));
```
Comment (qiaojialin) [inline example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteAlignedWithTSRecord.java:49]: As User could only see UnaryMeasurementSchema, we could rename this to MeasurementSchema

File: example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteAlignedWithTablet.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+      List<UnaryMeasurementSchema> measurementSchemas = new ArrayList<>();
+      measurementSchemas.add(new UnaryMeasurementSchema("s1", TSDataType.TEXT, TSEncoding.PLAIN));
+      measurementSchemas.add(new UnaryMeasurementSchema("s2", TSDataType.TEXT, TSEncoding.PLAIN));
+      measurementSchemas.add(new UnaryMeasurementSchema("s3", TSDataType.TEXT, TSEncoding.PLAIN));
+      measurementSchemas.add(new UnaryMeasurementSchema("s4", TSDataType.INT64, TSEncoding.RLE));
+
+      // register align timeseries
+      tsFileWriter.registerAlignedTimeseries(new Path(deviceId), measurementSchemas);
+
+      List<IMeasurementSchema> writeMeasurementScheams = new ArrayList<>();
```
Comment (qiaojialin) [inline example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteAlignedWithTablet.java:59]: Registering N and write M<N is a particular case, more cases is M=N.
So better to give an example that just use the measurementSchemas to write data.

File: example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteAlignedWithTablet.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      }
+    }
+    // write
+    if (tablet.rowSize != 0) {
+      tsFileWriter.writeAligned(tablet);
+      tablet.reset();
+    }
+  }
+
+  private static void writeWithTablet(TsFileWriter tsFileWriter)
```
Comment (qiaojialin) [inline example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteAlignedWithTablet.java:114]:
```suggestion
  private static void writeNonalignedWithTablet(TsFileWriter tsFileWriter)
```

File: example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteAlignedWithTablet.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    for (int r = 0; r < rowNum; r++, value++) {
+      int row = tablet.rowSize++;
+      timestamps[row] = timestamp++;
+      for (int i = 0; i < sensorNum; i++) {
+        long[] sensor = (long[]) values[i];
+        sensor[row] = value;
+      }
+      // write
+      if (tablet.rowSize == tablet.getMaxRowNumber()) {
+        tsFileWriter.write(tablet);
```
Comment (qiaojialin) [inline example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteAlignedWithTablet.java:141]:
```suggestion
        tsFileWriter.writeNonAligned(tablet);
```
Make some distinguish

File: example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteWithTSRecord.java [hunk 5]
```diff
@@ -39,43 +44,58 @@
...
-          tsRecord.addTuple(dPoint3);
-          // write TSRecord
-          tsFileWriter.write(tsRecord);
-        }
+        // register timeseries
+        tsFileWriter.registerTimeseries(new Path("root.sg.d1"), schemas.get(0));
+        tsFileWriter.registerTimeseries(new Path("root.sg.d1"), schemas.get(1));
+        tsFileWriter.registerTimeseries(new Path("root.sg.d1"), schemas.get(2));
+
+        List<IMeasurementSchema> writeMeasurementScheams = new ArrayList<>();
```
Comment (qiaojialin) [inline example/tsfile/src/main/java/org/apache/iotdb/tsfile/TsFileWriteWithTSRecord.java:64]: the same, give a most popular example.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/header/PageHeader.java [hunk 6]
```diff
@@ -54,6 +54,9 @@ public static int estimateMaxPageHeaderSizeWithoutStatistics() {
   public static PageHeader deserializeFrom(
       InputStream inputStream, TSDataType dataType, boolean hasStatistic) throws IOException {
     int uncompressedSize = ReadWriteForEncodingUtils.readUnsignedVarInt(inputStream);
+    if (uncompressedSize == 0) { // Empty Page
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/header/PageHeader.java:57]: when we get an empty page?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/reader/page/TimePageReader.java [hunk 7]
```diff
@@ -53,12 +54,15 @@ public TimePageReader(PageHeader pageHeader, ByteBuffer pageData, Decoder timeDe
   }
 
   public long[] nexTimeBatch() throws IOException {
-    long[] timeBatch = new long[(int) pageHeader.getStatistics().getCount()];
-    int index = 0;
+    List<Long> timeList = new ArrayList<>();
+    // long[] timeBatch = new long[(int) pageHeader.getStatistics().getCount()]; //
+    // Todo:bug.当TimePage有多页，statistics就会为null
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/reader/page/TimePageReader.java:59]: need to remove these lines

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkGroupWriterImpl.java [hunk 8]
```diff
@@ -45,150 +41,98 @@
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkGroupWriterImpl.java:41]: Rename to NonAlignedChunkGroupWriterImpl

**Review by qiaojialin — CHANGES_REQUESTED** [review]

**Comment by qiaojialin** [comment]
add UserGuide for TsFile API

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/header/ChunkHeader.java [hunk 9]
```diff
@@ -153,6 +153,8 @@ public static ChunkHeader deserializeFrom(InputStream inputStream, byte chunkTyp
       throws IOException {
     // read measurementID
     String measurementID = ReadWriteIOUtils.readVarIntString(inputStream);
+    measurementID =
+        measurementID == null ? "" : measurementID; // measurementID in TimeChunk header is null
```
Comment (wangchao316) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/header/ChunkHeader.java:157]: 1. Comments should be placed on the top line of code
2. Why is this parameter set to "" when measurementId is empty?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/reader/page/TimePageReader.java [hunk 10]
```diff
@@ -53,12 +54,11 @@ public TimePageReader(PageHeader pageHeader, ByteBuffer pageData, Decoder timeDe
...
   public long[] nexTimeBatch() throws IOException {
-    long[] timeBatch = new long[(int) pageHeader.getStatistics().getCount()];
-    int index = 0;
+    List<Long> timeList = new ArrayList<>();
     while (timeDecoder.hasNext(timeBuffer)) {
-      timeBatch[index++] = timeDecoder.readLong(timeBuffer);
+      timeList.add(timeDecoder.readLong(timeBuffer));
     }
-    return timeBatch;
+    return timeList.stream().mapToLong(t -> t.longValue()).toArray();
```
Comment (wangchao316) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/reader/page/TimePageReader.java:61]: Why is this modification necessary? This modification will cause performance loss, which is equivalent to two iterations. o(n*n)

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/AlignedChunkGroupWriterImpl.java [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+      if (checkIsHistoryData("", time)) {
+        continue;
+      }
+      for (int columnIndex = 0; columnIndex < measurementSchemas.size(); columnIndex++) {
+        writenMeasurementSet.add(measurementSchemas.get(columnIndex).getMeasurementId());
+        boolean isNull = false;
+        // check isNull by bitMap in tablet
+        if (tablet.bitMaps != null
+            && tablet.bitMaps[columnIndex] != null
+            && tablet.bitMaps[columnIndex].isMarked(row)) {
```
Comment (JackieTien97) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/AlignedChunkGroupWriterImpl.java:159]:
```suggestion
            && !tablet.bitMaps[columnIndex].isMarked(row)) {
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkGroupWriterImpl.java [hunk 12]
```diff
@@ -45,150 +41,98 @@
...
+        hasOneColumnWritten = true;
         boolean isNull = false;
         // check isNull by bitMap in tablet
         if (tablet.bitMaps != null
-            && tablet.bitMaps[columnIndex] != null
-            && tablet.bitMaps[columnIndex].isMarked(row)) {
+            && tablet.bitMaps[column] != null
+            && tablet.bitMaps[column].isMarked(row)) {
           isNull = true;
         }
```
Comment (JackieTien97) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkGroupWriterImpl.java:105]: `isNUll` is useless

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java [hunk 13]
```diff
@@ -457,4 +457,8 @@ public boolean isMerging() {
   public void setLastPoint(boolean isLastPoint) {
     this.isLastPoint = isLastPoint;
   }
+
+  public void setValueIndex(int valueIndex) {
+    throw new IllegalStateException("write time method is not implemented in common chunk writer");
+  }
```
Comment (JackieTien97) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java]: This method is useless any more.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java [hunk 14]
```diff
@@ -328,7 +328,7 @@ public long estimateMaxSeriesMemSize() {
 
   @Override
   public long getSerializedChunkSize() {
-    if (pageBuffer.size() == 0) {
+    if (pageBuffer.size() == 0 || statistics.getCount() == 0) {
```
Comment (JackieTien97) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java:331]: In which case, `pageBuffer.size() != 0` and `statistics.getCount() == 0`?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ValueChunkWriter.java [hunk 15]
```diff
@@ -180,14 +203,41 @@ public long estimateMaxSeriesMemSize() {
   }
 
   public long getCurrentChunkSize() {
-    if (pageBuffer.size() == 0) {
+    if (pageBuffer.size() == 0 || statistics.getCount() == 0) {
```
Comment (JackieTien97) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ValueChunkWriter.java:206]: Even if `statistics.getCount() == 0`, the chunk size is not 0.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/TsFileWriter.java [hunk 16]
```diff
@@ -151,92 +189,205 @@ protected TsFileWriter(TsFileIOWriter fileWriter, Schema schema, TSFileConfig co
...
-  public void registerTimeseries(Path path, IMeasurementSchema measurementSchema)
-      throws WriteProcessException {
-    if (schema.containsTimeseries(path)) {
-      throw new WriteProcessException("given timeseries has exists! " + path);
+  public void registerTimeseries(Path devicePath, List<UnaryMeasurementSchema> measurementSchemas) {
+    for (UnaryMeasurementSchema schema : measurementSchemas) {
+      try {
+        registerTimeseries(devicePath, schema);
+      } catch (WriteProcessException e) {
+        LOG.error(e.getMessage());
```
Comment (wangchao316) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/TsFileWriter.java:214]: if exception does not deal,  please print warn log.

**Review by wangchao316 — CHANGES_REQUESTED** [review]

**Review by wangchao316 — COMMENTED** [review]
I suggest should use benchmark for Stress test.  Check whether the performance deteriorates compared with the previous one.

**Pull request merged**